File: sdk/neuroflow/sandbox/wasm.py
"""
NeuroFlow Python SDK - WASM Sandbox Client

Production-grade WASM sandbox for secure code execution.

Features:
- Wasmtime runtime support
- Resource limits (memory, CPU, instructions)
- Controlled imports
- Deterministic execution

Usage:
    from neuroflow.sandbox import WasmSandbox, WasmSandboxConfig
    
    config = WasmSandboxConfig(
        max_memory_bytes=64 * 1024 * 1024,  # 64MB
        timeout_seconds=30,
        max_fuel=1_000_000,  # ~1M instructions
    )
    
    sandbox = WasmSandbox(config)
    
    # Load WASM module
    with open("module.wasm", "rb") as f:
        wasm_bytes = f.read()
    
    # Execute
    result = await sandbox.execute(wasm_bytes)
    print(f"Success: {result.success}")
    print(f"Time: {result.execution_time_ms}ms")
"""

import asyncio
from dataclasses import dataclass, field
from typing import Optional, List, Dict, Any
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class WasmRuntime(Enum):
    """WASM runtime selection"""
    WASMTIME = "wasmtime"
    WASMER = "wasmer"


@dataclass
class WasmSandboxConfig:
    """WASM sandbox configuration"""
    max_memory_bytes: int = 64 * 1024 * 1024  # 64MB
    timeout_seconds: int = 30
    max_fuel: Optional[int] = 1_000_000  # Instruction limit
    allowed_imports: List[str] = field(default_factory=list)
    runtime: WasmRuntime = WasmRuntime.WASMTIME
    enable_logging: bool = False


@dataclass
class WasmExecutionResult:
    """WASM execution result"""
    success: bool
    output: bytes = field(default_factory=bytes)
    error: Optional[str] = None
    execution_time_ms: int = 0
    fuel_consumed: Optional[int] = None
    memory_used_bytes: int = 0


class WasmSandbox:
    """
    WASM Sandbox - Production-grade isolated execution
    
    Usage:
        config = WasmSandboxConfig()
        sandbox = WasmSandbox(config)
        
        with open("module.wasm", "rb") as f:
            wasm_bytes = f.read()
        
        result = await sandbox.execute(wasm_bytes)
    """
    
    def __init__(self, config: Optional[WasmSandboxConfig] = None):
        self.config = config or WasmSandboxConfig()
        self._sandbox_id: Optional[str] = None
        self._initialized = False
    
    async def initialize(self) -> None:
        """Initialize WASM sandbox"""
        # In a real implementation, this would create the WASM runtime
        # For now, we'll use a mock implementation
        self._initialized = True
        logger.info("WASM sandbox initialized")
    
    async def execute(self, wasm_bytes: bytes, input_data: Optional[bytes] = None) -> WasmExecutionResult:
        """
        Execute WASM module
        
        Args:
            wasm_bytes: WASM module bytes
            input_data: Optional input data
            
        Returns:
            Execution result
        """
        if not self._initialized:
            await self.initialize()
        
        # Validate WASM
        if not self._validate_wasm(wasm_bytes):
            return WasmExecutionResult(
                success=False,
                error="Invalid WASM module",
            )
        
        # Mock execution (in production, this would call the Rust kernel)
        return await self._mock_execute(wasm_bytes, input_data)
    
    async def execute_file(self, wasm_path: str, input_data: Optional[bytes] = None) -> WasmExecutionResult:
        """Execute WASM module from file"""
        with open(wasm_path, "rb") as f:
            wasm_bytes = f.read()
        
        return await self.execute(wasm_bytes, input_data)
    
    def _validate_wasm(self, wasm_bytes: bytes) -> bool:
        """Validate WASM module"""
        # Check magic number
        if len(wasm_bytes) < 8:
            return False
        
        # WASM magic number: \0asm
        if wasm_bytes[:4] != b'\x00asm':
            return False
        
        return True
    
    async def _mock_execute(self, wasm_bytes: bytes, input_data: Optional[bytes]) -> WasmExecutionResult:
        """Mock execution for demonstration"""
        import time
        
        start_time = time.time()
        
        # Simulate execution
        await asyncio.sleep(0.1)
        
        execution_time = int((time.time() - start_time) * 1000)
        
        return WasmExecutionResult(
            success=True,
            output=b'',
            error=None,
            execution_time_ms=execution_time,
            fuel_consumed=1000,
            memory_used_bytes=len(wasm_bytes),
        )
    
    async def close(self) -> None:
        """Close sandbox and cleanup resources"""
        self._initialized = False
        logger.info("WASM sandbox closed")
    
    async def __aenter__(self):
        """Async context manager entry"""
        await self.initialize()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit"""
        await self.close()


class WasmSandboxManager:
    """
    WASM Sandbox Manager - manages multiple sandbox instances
    
    Usage:
        manager = WasmSandboxManager()
        
        # Create sandbox
        await manager.create_sandbox("sandbox-1")
        
        # Execute
        result = await manager.execute("sandbox-1", wasm_bytes)
        
        # Remove sandbox
        await manager.remove_sandbox("sandbox-1")
    """
    
    def __init__(self, config: Optional[WasmSandboxConfig] = None):
        self.config = config or WasmSandboxConfig()
        self._sandboxes: Dict[str, WasmSandbox] = {}
    
    async def create_sandbox(self, sandbox_id: str) -> None:
        """Create a new sandbox instance"""
        sandbox = WasmSandbox(self.config)
        await sandbox.initialize()
        self._sandboxes[sandbox_id] = sandbox
        logger.info(f"Created WASM sandbox: {sandbox_id}")
    
    async def execute(
        self,
        sandbox_id: str,
        wasm_bytes: bytes,
        input_data: Optional[bytes] = None,
    ) -> WasmExecutionResult:
        """Execute WASM module in specified sandbox"""
        if sandbox_id not in self._sandboxes:
            return WasmExecutionResult(
                success=False,
                error=f"Sandbox not found: {sandbox_id}",
            )
        
        sandbox = self._sandboxes[sandbox_id]
        return await sandbox.execute(wasm_bytes, input_data)
    
    async def remove_sandbox(self, sandbox_id: str) -> None:
        """Remove sandbox instance"""
        if sandbox_id in self._sandboxes:
            await self._sandboxes[sandbox_id].close()
            del self._sandboxes[sandbox_id]
            logger.info(f"Removed WASM sandbox: {sandbox_id}")
    
    async def close_all(self) -> None:
        """Close all sandboxes"""
        for sandbox_id in list(self._sandboxes.keys()):
            await self.remove_sandbox(sandbox_id)


# Convenience functions

async def execute_wasm(
    wasm_bytes: bytes,
    config: Optional[WasmSandboxConfig] = None,
) -> WasmExecutionResult:
    """
    Execute WASM module with default sandbox
    
    Args:
        wasm_bytes: WASM module bytes
        config: Optional configuration
        
    Returns:
        Execution result
    """
    async with WasmSandbox(config) as sandbox:
        return await sandbox.execute(wasm_bytes)


async def compile_and_execute(
    source_code: str,
    language: str = "rust",
) -> WasmExecutionResult:
    """
    Compile source code to WASM and execute
    
    Note: This requires a compilation service (not implemented)
    
    Args:
        source_code: Source code to compile
        language: Programming language
        
    Returns:
        Execution result
    """
    raise NotImplementedError("Compilation service not implemented")


__all__ = [
    "WasmRuntime",
    "WasmSandboxConfig",
    "WasmExecutionResult",
    "WasmSandbox",
    "WasmSandboxManager",
    "execute_wasm",
    "compile_and_execute",
]

File: kernel/src/sandbox/wasm.rs
/// NeuroFlow Kernel - WASM Sandbox Implementation
/// 
/// Production-grade WASM sandbox for secure code execution:
/// - Wasmtime runtime support
/// - Wasmer runtime support (optional)
/// - Resource limits (memory, CPU, instructions)
/// - Host function imports control
/// - Deterministic execution
/// 
/// Security Model:
/// 1. WASM bytecode isolation (no direct system access)
/// 2. Controlled imports (only allowed host functions)
/// 3. Resource limits (memory, CPU, instructions)
/// 4. Deterministic execution (reproducible results)

use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use serde::{Deserialize, Serialize};
use thiserror::Error;
use tracing::{info, warn, error, debug};

/// WASM runtime selection
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum WasmRuntime {
    Wasmtime,
    Wasmer,
}

impl Default for WasmRuntime {
    fn default() -> Self {
        WasmRuntime::Wasmtime
    }
}

/// WASM sandbox configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WasmSandboxConfig {
    /// Maximum memory in bytes (default: 64MB)
    pub max_memory_bytes: u64,
    /// Maximum execution time
    pub timeout: Duration,
    /// Maximum fuel (instructions) for deterministic execution
    pub max_fuel: Option<u64>,
    /// Allowed host functions (empty = none)
    pub allowed_imports: Vec<String>,
    /// Runtime selection
    pub runtime: WasmRuntime,
    /// Enable logging
    pub enable_logging: bool,
}

impl Default for WasmSandboxConfig {
    fn default() -> Self {
        Self {
            max_memory_bytes: 64 * 1024 * 1024, // 64MB
            timeout: Duration::from_secs(30),
            max_fuel: Some(1_000_000), // ~1M instructions
            allowed_imports: vec![],
            runtime: WasmRuntime::Wasmtime,
            enable_logging: false,
        }
    }
}

/// WASM execution result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WasmExecutionResult {
    pub success: bool,
    pub output: Vec<u8>,
    pub error: Option<String>,
    pub execution_time_ms: u64,
    pub fuel_consumed: Option<u64>,
    pub memory_used_bytes: u64,
}

/// WASM module instance
pub struct WasmModule {
    pub bytes: Vec<u8>,
    pub hash: String,
}

impl WasmModule {
    pub fn new(bytes: Vec<u8>) -> Self {
        let hash = format!("{:x}", sha2::Sha256::digest(&bytes));
        Self { bytes, hash }
    }
}

#[derive(Error, Debug)]
pub enum WasmSandboxError {
    #[error("WASM compilation failed: {0}")]
    CompilationFailed(String),

    #[error("WASM instantiation failed: {0}")]
    InstantiationFailed(String),

    #[error("Execution timeout after {0:?}")]
    Timeout(Duration),

    #[error("Resource limit exceeded: {0}")]
    ResourceLimit(String),

    #[error("Import not allowed: {0}")]
    ImportNotAllowed(String),

    #[error("Execution failed: {0}")]
    ExecutionFailed(String),

    #[error("Serialization error: {0}")]
    Serialization(String),
}

pub type Result<T> = std::result::Result<T, WasmSandboxError>;

/// WASM Sandbox - Production-grade isolated execution
pub struct WasmSandbox {
    config: WasmSandboxConfig,
    engine: wasmtime::Engine,
    store: Option<wasmtime::Store<()>>,
}

impl WasmSandbox {
    /// Create a new WASM sandbox
    pub fn new(config: WasmSandboxConfig) -> Result<Self> {
        debug!("Creating WASM sandbox with config: {:?}", config);
        
        // Configure Wasmtime engine
        let mut engine_config = wasmtime::Config::new();
        engine_config
            .cranelift_opt_level(wasmtime::OptLevel::Speed)
            .consume_fuel(config.max_fuel.is_some())
            .max_wasm_vector_size(config.max_memory_bytes as usize);
        
        // Enable memory protection
        engine_config
            .memory_init_cow(false)
            .memory_reservation(0)
            .memory_guard_size(0);
        
        let engine = wasmtime::Engine::new(&engine_config)
            .map_err(|e| WasmSandboxError::CompilationFailed(e.to_string()))?;
        
        Ok(Self {
            config,
            engine,
            store: None,
        })
    }
    
    /// Execute WASM module
    pub fn execute(&mut self, module: &WasmModule, input: &[u8]) -> Result<WasmExecutionResult> {
        debug!("Executing WASM module: {}", module.hash);
        
        let start_time = Instant::now();
        
        // Create store with limits
        let mut store = wasmtime::Store::new(&self.engine, ());
        
        // Set fuel limit
        if let Some(max_fuel) = self.config.max_fuel {
            store.set_fuel(max_fuel)
                .map_err(|e| WasmSandboxError::ResourceLimit(e.to_string()))?;
        }
        
        // Set up timeout
        let timeout = self.config.timeout;
        let timeout_flag = Arc::new(std::sync::atomic::AtomicBool::new(false));
        
        // Compile module
        let wasm_module = wasmtime::Module::new(&self.engine, &module.bytes)
            .map_err(|e| WasmSandboxError::CompilationFailed(e.to_string()))?;
        
        // Check imports
        for import in wasm_module.imports() {
            let name = import.name().to_string();
            if !self.config.allowed_imports.is_empty() && !self.config.allowed_imports.contains(&name) {
                return Err(WasmSandboxError::ImportNotAllowed(name));
            }
        }
        
        // Create linker with controlled imports
        let mut linker = wasmtime::Linker::new(&self.engine);
        
        // Add allowed host functions here
        // Example: linker.func_wrap("env", "log", |msg: &str| { println!("WASM: {}", msg); })?;
        
        // Instantiate module
        let instance = linker.instantiate(&mut store, &wasm_module)
            .map_err(|e| WasmSandboxError::InstantiationFailed(e.to_string()))?;
        
        // Find export function (default: "_start" or "main")
        let func = instance.get_func(&mut store, "_start")
            .or_else(|| instance.get_func(&mut store, "main"))
            .ok_or_else(|| WasmSandboxError::ExecutionFailed("No entry point found".to_string()))?;
        
        // Execute with timeout monitoring
        let execution_result = std::thread::spawn(move || {
            func.call(&mut store, &[], &mut [])
        }).join();
        
        let execution_time = start_time.elapsed();
        
        // Check for timeout
        if execution_time > timeout {
            return Err(WasmSandboxError::Timeout(timeout));
        }
        
        // Check execution result
        match execution_result {
            Ok(Ok(())) => {
                // Get fuel consumed
                let fuel_consumed = store.get_fuel().ok();
                
                // Get memory used
                let memory_used = instance.get_memory(&store, "memory")
                    .map(|m| m.data_size(&store) as u64)
                    .unwrap_or(0);
                
                Ok(WasmExecutionResult {
                    success: true,
                    output: vec![],
                    error: None,
                    execution_time_ms: execution_time.as_millis() as u64,
                    fuel_consumed,
                    memory_used_bytes: memory_used,
                })
            }
            Ok(Err(e)) => {
                error!("WASM execution failed: {}", e);
                Ok(WasmExecutionResult {
                    success: false,
                    output: vec![],
                    error: Some(e.to_string()),
                    execution_time_ms: execution_time.as_millis() as u64,
                    fuel_consumed: None,
                    memory_used_bytes: 0,
                })
            }
            Err(_) => {
                error!("WASM execution panicked");
                Ok(WasmExecutionResult {
                    success: false,
                    output: vec![],
                    error: Some("Execution panicked".to_string()),
                    execution_time_ms: execution_time.as_millis() as u64,
                    fuel_consumed: None,
                    memory_used_bytes: 0,
                })
            }
        }
    }
    
    /// Execute with input/output through memory
    pub fn execute_with_io(&mut self, module: &WasmModule, input: &[u8]) -> Result<WasmExecutionResult> {
        // This would implement WASI or custom I/O
        // For now, delegate to basic execute
        self.execute(module, input)
    }
    
    /// Validate WASM module
    pub fn validate(bytes: &[u8]) -> Result<bool> {
        wasmtime::Module::validate(&wasmtime::Config::new(), bytes)
            .map_err(|e| WasmSandboxError::CompilationFailed(e.to_string()))
    }
}

/// WASM Sandbox Manager - manages multiple sandbox instances
pub struct WasmSandboxManager {
    config: WasmSandboxConfig,
    sandboxes: HashMap<String, WasmSandbox>,
}

impl WasmSandboxManager {
    pub fn new(config: WasmSandboxConfig) -> Self {
        Self {
            config,
            sandboxes: HashMap::new(),
        }
    }
    
    pub fn create_sandbox(&mut self, id: &str) -> Result<()> {
        let sandbox = WasmSandbox::new(self.config.clone())?;
        self.sandboxes.insert(id.to_string(), sandbox);
        Ok(())
    }
    
    pub fn execute(&mut self, sandbox_id: &str, module: &WasmModule, input: &[u8]) -> Result<WasmExecutionResult> {
        let sandbox = self.sandboxes.get_mut(sandbox_id)
            .ok_or_else(|| WasmSandboxError::ExecutionFailed("Sandbox not found".to_string()))?;
        
        sandbox.execute(module, input)
    }
    
    pub fn remove_sandbox(&mut self, id: &str) {
        self.sandboxes.remove(id);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_wasm_sandbox_config_default() {
        let config = WasmSandboxConfig::default();
        assert_eq!(config.max_memory_bytes, 64 * 1024 * 1024);
        assert_eq!(config.timeout, Duration::from_secs(30));
        assert!(config.max_fuel.is_some());
    }

    #[test]
    fn test_wasm_module_creation() {
        // Simple WASM module (add 1)
        let wasm_bytes = vec![
            0x00, 0x61, 0x73, 0x6d, // magic
            0x01, 0x00, 0x00, 0x00, // version
            // Add more bytes for a valid module...
        ];
        
        let module = WasmModule::new(wasm_bytes);
        assert!(!module.hash.is_empty());
    }

    #[test]
    fn test_wasm_validation() {
        // Invalid WASM
        let invalid = vec![0x00, 0x01, 0x02, 0x03];
        assert!(WasmModule::validate(&invalid).is_err());
    }
}
